jobs/haproxy/templates/drain.erb

```diff
@@ -6,13 +6,21 @@ set -e
 pidfile=/var/vcap/sys/run/bpm/haproxy/haproxy.pid
 sockfile=/var/vcap/sys/run/haproxy/stats.sock
 logfile=/var/vcap/sys/log/haproxy/drain.log
+lockfile=/var/vcap/sys/run/haproxy/drain.lock
+
+mkdir -p "$(dirname ${logfile})"
 
 <% if not p("ha_proxy.drain_enable") -%>
 echo "drain is disabled" >> ${logfile}
 echo 0
 exit 0
 <% else -%>
-mkdir -p "$(dirname ${logfile})"
+
+if [[ -f ${lockfile} ]]; then
+  echo "$(date): draining already in progress" >> ${logfile}
+  echo 0
+  exit 0
+fi
```

Comment:
the lock exists to decouple draining from restarts. the file will be removed during `haproxy_wrapper` startup in the `cleanup_daemon` function together with the rest (haproxy stats socket etc.)
we could think about adding a timestamp check here, but it's hard to find the right time out value:
- the time for draining will be at least grace_time + drain_time
- then there will be a potentially indefinite time of pre-stop scripts running
- then we have a potential bpm timeout of 15s to actually stop haproxy_wrapper
- then we have the pre-start script where we (potentially) wait for additional external certificates before we actually start
- finally, the time before haproxy has come online again, healthy and with the socket ready to accept commands, until we potentially can start a new drain again

The only possible way right now, this could get stuck is if the wrapper_script fails to start and fails to remove the lock file again. If the drain script starts draining but fails somewhere in-between, the next try will simply skip draining because the lock is there and stop haproxy without draining, clearing the file again and making it unstuck.